core/ledger/kvledger/txmgmt/statedb/statecouchdb/statecouchdb.go

```diff
@@ -197,6 +202,21 @@ func newVersionedDB(couchInstance *couchInstance, redoLogger *redoLogger, dbName
 		redoLogger:         redoLogger,
 		cache:              cache,
 	}
+
+	vdb.channelMetadata, err = vdb.getSavedChannelMetadata()
+	if err != nil {
+		return nil, err
+	}
+	if vdb.channelMetadata == nil {
+		vdb.channelMetadata = &channelMetadata{
+			ChannelName:      chainName,
+			NamespaceDBsInfo: make(map[string]*namespaceDBInfo),
+		}
+		if err = vdb.updateAndSaveNamespaceDBsInfo("", dbName); err != nil {
```

Comment:
This could lead to missing a mapping, as we allow the empty namespace as well (we write channel config in that). Do we really need to store the mapping for the channel metadata itself? We anyways, always need to know it's name in order to get in to pull out the mapping.

Reply:
I don't understand why this could lead to missing a mapping. The empty namespace is same as metadataDB, which will not be missing since it has been added here. Agree we don't really need to store the metadataDB mapping. However, even if it is not added here, it will be added later when getNamespaceDBHandle is called with empty namespace.

Reply:
Unfortunately, there are two different functions for generating the names.
- For the channel metadatadb - [constructMetadataDBName](https://github.com/hyperledger/fabric/blob/e461141a2dfaa1214de60a47f3114c8b8b4805ef/core/ledger/kvledger/txmgmt/statedb/statecouchdb/couchdbutil.go#L186) 
- For the user namespace db - [constructNamespaceDBName](https://github.com/hyperledger/fabric/blob/e461141a2dfaa1214de60a47f3114c8b8b4805ef/core/ledger/kvledger/txmgmt/statedb/statecouchdb/couchdbutil.go#L205)
Logically, they can produce different outputs. So, if you always populate the entry in the map with an empty namespace with the output of first function, what would happen when you all the second function with an empty namespace?

Reply:
The introduction of `channelName string` indirectly stores the mapping and hence, we don't need to store it explicitly. Though there won't be any collusion with the channelconfig namespace here, it adds a bit of confusion due to the usage of an empty namespace for metadataDBName. Further, the `getNamespaceDBHandle()` is not involved when accessing metadataDB.  I think the storage of metadataDBName mapping is forced here as the `updateAndSaveNamespaceDBsInfo()` has the following code
```
	if _, ok := vdb.channelMetadata.NamespaceDBsInfo[ns]; ok {
 		return nil
 	}

 	logger.Debugf("[%s] add namespaceDBInfo for namespace %s", vdb.chainName, ns)
 	vdb.channelMetadata.NamespaceDBsInfo[ns] = &namespaceDBInfo{
 		Namespace: ns,
 		DBName:    dbName,
 	}
```
If these line can be moved to `getNamespaceDBHandle()`, we can rename `updateAndSaveNamespaceDBsInfo()`  to `commitChannelMetadata()` and skip storing the mapping between metadatDBName and channel explicitly (given that channelName stores it implicitly). 

Otherwise, we need to avoid this confusion using a comment if you are not comfortable in modifying the `updateAndSaveNamespaceDBsInfo()`

Reply:
Ah. Correcting myself. A collision does happen on the namespace but not on the dbName.

Reply:
Thank you @manish-sethi and @cendhu. Will remove metadataDB from mapping. 
A side note for metadataDB and namespaceDB. Based on existing implementation, there is dbName collision between metadataDB and namespaceDB when namespace is an empty string. However if channel name is too long and truncated, their dbNames are not same any more.

Reply:
I see. Nice find @wenjianqiao So the channelconfig block could be either in the metadataDB or in a separate DB based on the length of the channel name. We may have to live with this non-deterministic behaviour for now. Please add a comment stating this behaviour and create a JIRA task too. As part of upgradeDB or rebuildDB commands, we can think of addressing this issue. We can modify [constructMetadataDBName](https://github.com/hyperledger/fabric/blob/e461141a2dfaa1214de60a47f3114c8b8b4805ef/core/ledger/kvledger/txmgmt/statedb/statecouchdb/couchdbutil.go#L186) to return a name that cannot collide with empty namespace's dbName  and migrate the content from existing metadataDB to newMetadataDB in the `newVersionedDB() `